src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspace.cs

```diff
@@ -114,5 +114,19 @@ public PortableExecutableReference CreatePortableExecutableReference(string file
         }
 
         internal abstract string TryGetRuleSetPathForProject(ProjectId projectId);
+
+        /// <summary>
+        /// Attempts to sync a named type name and namespace based on file path. PreviousFilePath is 
+        /// used to determine how the policy might be applied, assuming that matching types with previous
+        /// indicates a user wants the type to match with the new location or name.
+        /// May fail if:
+        /// - The appropriate type for naming policy cannot be found
+        /// - The type is partial and spread across multiple files
+        /// - The solution changes during the operation
+        /// </summary>
+        /// <returns> 
+        /// Returns false if the symbol update failed, true if it succeeded
+        /// </returns>
+        public abstract Task<bool> TryUpdateNamedTypeToMatchFilePath(DocumentId documentId, string previousFilePath, CancellationToken cancellation = CancellationToken.None);
```

Comment:
Do we want to explicitly say this is updating the named type? Or be a bit more vague about "UpdateFileContentsAfterFileRename"? (see my earlier comment about maybe updating file headers)

Reply:
I wanted to avoid "Rename" terminology to leave it open ended for things like file addition or movement. While moving a file _is_ a rename as far as full path is concerned, it's harder to convey that information. Maybe if we move to a service it could be `IDocumentContentPolicyEnforcer` (terrible name, but gets my point across) 

```c# 
public interface IDocumentContentPolicyEnforcer
{
    Task<bool> TryApplyPolicy(DocumentId documentId, Document oldDocument, CancellationToken cancellationToken = CancellationToken.None) 
}
```

Reply:
With current design, I think we can keep it specifically named. The service approach allows more context in calling, and if we want to add an overarching `UpdateFileContentsAfterFileRename` that does multiple things we can.